src/ethereum_test_specs/base_static.py

```diff
@@ -0,0 +1,133 @@
+"""Base class to parse test cases written in static formats."""
+
+from abc import abstractmethod
+from typing import Any, Callable, ClassVar, Dict, List, Type, Union
+
+from pydantic import BaseModel, TypeAdapter, ValidatorFunctionWrapHandler, model_validator
+
+
+class BaseStaticTest(BaseModel):
+    """Represents a base class that reads cases from static files."""
+
+    formats: ClassVar[List[Type["BaseStaticTest"]]] = []
+    formats_type_adapter: ClassVar[TypeAdapter]
+
+    format_name: ClassVar[str] = ""
+
+    @classmethod
+    def __pydantic_init_subclass__(cls, **kwargs):
+        """
+        Register all subclasses of BaseStaticTest with a static test format name set
+        as possible static test format.
+        """
+        if cls.format_name:
+            # Register the new fixture format
+            BaseStaticTest.formats.append(cls)
+            if len(BaseStaticTest.formats) > 1:
+                BaseStaticTest.formats_type_adapter = TypeAdapter(
+                    Union[tuple(BaseStaticTest.formats)],
+                )
+            else:
+                BaseStaticTest.formats_type_adapter = TypeAdapter(cls)
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def _parse_into_subclass(
+        cls, v: Any, handler: ValidatorFunctionWrapHandler
+    ) -> "BaseStaticTest":
+        """Parse the static test into the correct subclass."""
+        if cls is BaseStaticTest:
+            return BaseStaticTest.formats_type_adapter.validate_python(v)
+        return handler(v)
+
+    @abstractmethod
+    def fill_function(self) -> Callable:
+        """
+        Return the test function that can be used to fill the test.
+
+        This method should be implemented by the subclasses.
+
+        The function returned can be optionally decorated with the `@pytest.mark.parametrize`
+        decorator to parametrize the test with the number of sub test cases.
+
+        Example:
+        ```
+        @pytest.mark.parametrize("n", [1])
+        @pytest.mark.parametrize("m", [1, 2])
+        @pytest.mark.valid_from("Homestead")
+        def test_state_filler(
+            state_test: StateTestFiller,
+            fork: Fork,
+            pre: Alloc,
+            n: int,
+            m: int,
+        ):
+            \"\"\"Generate a test from a static state filler.\"\"\"
+            assert n == 1
+            assert m in [1, 2]
+            env = Environment(**self.env.model_dump())
+            sender = pre.fund_eoa()
+            tx = Transaction(
+                ty=0x0,
+                nonce=0,
+                to=Address(0x1000),
+                gas_limit=500000,
+                protected=False if fork in [Frontier, Homestead] else True,
+                data="",
+                sender=sender,
+            )
+            state_test(env=env, pre=pre, post={}, tx=tx)
+
+        return test_state_filler
```

Comment:
the function return itself?

Reply:
Yes, the returned function is dynamically generated based on the properties of the specific static test that was read from the file.